DevOp For Sphinx Documentation

engrosamaali91 opened the pull request:
The branch DevOp has been rebased on dev branch.

> Steps followed:

  1. git checkout dev
  2. git pull upstream dev
  3. git checkout DevOp
  4. git rebase dev
  5. git push origin DevOp --force

engrosamaali91 committed:
modified numpy docstring continuation

engrosamaali91 committed:
Missing docstrings added in new files

engrosamaali91 committed:
recon_surf directory

engrosamaali91 committed:
sphinx doc files

engrosamaali91 committed:
DevOp rebased on dev

dkuegler commented on FastSurferCNN/data_loader/data_utils.py:
```diff
@@ -565,22 +566,22 @@ def deep_sulci_and_wm_strand_mask(
     ----------
     lut_file : str
-        path and name of FreeSurfer-style LUT file with classes of interest
+        Path and name of FreeSurfer-style LUT file with classes of interest
```
the point should be at the end of the line here not 3 lines below

engrosamaali91 replied:
PR09: Parameter "lut_file" description should finish with "."

The description of lut_file parameter must end with a period to avoid PRO9 warning.
It Is taking the example entry as part of the lut_file description.

dkuegler commented on FastSurferCNN/data_loader/data_utils.py:
```diff
@@ -1037,20 +1038,20 @@ def infer_mapping_from_lut(
 ) -> np.ndarray:
-    """[MISSING].
+    """
+    [MISSING].
```
Guess the mapping from a lookup table.

dkuegler commented on FastSurferCNN/data_loader/dataset.py:
```diff
@@ -483,19 +496,20 @@ def _get_scale_factor(self, img_zoom):
         ----------
         img_zoom :
-            zooming factor [MISSING]
+            Zooming factor [MISSING].
```
... / voxel sizes of the image.
type should be np.ndarray

dkuegler commented on FastSurferCNN/data_loader/dataset.py:
```diff
@@ -483,19 +496,20 @@ def _get_scale_factor(self, img_zoom):
         np.ndarray : float32
-            scale factor along x and y dimension
-
+            Scale factor along x and y dimension.
```
the type above is not a valid format. use numpy.typing.NDArray[float] or npt.NDArray[float]

dkuegler commented on FastSurferCNN/generate_hdf5.py:
```diff
@@ -99,35 +89,54 @@ class H5pyDataset:
-            patter (str): Pattern to match files in directory.
-        processing : str
-            Use aseg (Default value = "aparc")
+            - sag_mask (tuple[str, str]): Suffixes of labels names to mask for final sagittal labels.
```
is this change of the name also in the code?

dkuegler replied:
currently line 158 should be changed then and you need to do a search if this renaming also affects any other code outside of this method.

dkuegler commented on FastSurferCNN/generate_hdf5.py:
```diff
@@ -261,19 +272,18 @@ def _pad_image(self, img: npt.NDArray, max_out: int) -> np.ndarray:
         blt : int
-            Blank sliec threshold
-
+            Blank sliec threshold.
```
typo

dkuegler commented on FastSurferCNN/inference.py:
```diff
@@ -224,69 +224,122 @@ def load_checkpoint(self, ckpt: Union[str, os.PathLike]):
 
     def get_modelname(self):
-        """Return the model name."""
+        """
```
should probably also add the type to the declaration, i.e. `def get_modelname(self) -> str:`

Same for the methods below

dkuegler commented on FastSurferCNN/inference.py:
```diff
@@ -224,69 +224,122 @@ def load_checkpoint(self, ckpt: Union[str, os.PathLike]):
+
+        Returns
+        -------
+        Union[int, Tuple[int, int]]
```
I think numpy format also supports `int, tuple[int, int]` ; for the typing above, it should be `... -> int | tuple[int, int]`

also note the capitalization, tuple does not need the type from typing, you can now use the builtin `tuple`.

dkuegler commented on FastSurferCNN/reduce_to_aseg.py:
```diff
@@ -13,16 +13,17 @@
 # limitations under the License.
 
 
+import copy
```
the comment #IMPORTS should be first

dkuegler commented on FastSurferCNN/run_prediction.py:
```diff
@@ -140,44 +146,45 @@ def removesuffix(string: str, suffix: str) -> str:
+    pred_name : str
+    conf_name : str
+    orig_name : str
+    vox_size : Union[float, Literal["min"]]
```
vox_size: float, 'min'

Generally:
https://numpydoc.readthedocs.io/en/latest/format.html

For example:
```
Parameters
----------
filename : str
copy : bool
dtype : data-type
iterable : iterable object
shape : int or tuple of int
files : list of str
```

engrosamaali91 replied:
Already resolved :)

dkuegler commented on FastSurferCNN/run_prediction.py:
```diff
@@ -140,44 +146,45 @@ def removesuffix(string: str, suffix: str) -> str:
+    current_plane : str
+    models : Dict[str, Inference]
+    view_ops : Dict[str, Dict[str, Any]]
+    conform_to_1mm_threshold : Optional[float]
```
float, optional

dkuegler commented on FastSurferCNN/models/interpolation_layer.py:
```diff
@@ -186,29 +187,27 @@ def forward(
 
         Yields
         ------
         _T.Iterable[_T.Tuple[T_Scale, int]]
```
This is probably incorrect and should be `tuple of scale and int` or `tuple[T_Scale, int]`

dkuegler commented on FastSurferCNN/models/interpolation_layer.py:
```diff
@@ -366,39 +367,40 @@ def _calculate_crop_pad(
+    """
+    Perform a crop and interpolation on a Four-dimensional Tensor respecting batch and channel.
 
     Attributes
```
attributes starting with _ should be labeled as protected attributes or not be in the list at all, if there is a property for that specific variable, its fine, but else, these are protected attributes and should not appear in the "public" documentation.

same for methods, so label _interpolate as a protected helper function

engrosamaali91 replied:
Removed both private attributes and labeled _interpolate method as protected like this.    

class Zoom2d(_ZoomNd):

    """
    Perform a crop and interpolation on a Four-dimensional Tensor respecting batch and channel.

    Methods
    -------
    _interpolate
        (Protected) Crops, interpolates and pads the tensor.
    """

dkuegler commented on FastSurferCNN/models/interpolation_layer.py:
```diff
@@ -417,27 +419,27 @@ def __init__(
         self._crop_position = crop_position
```
the code here should be `self.crop_position = crop_position`  (e.g. crop_position should probably be added as a property with the code in lines 408ff + 419 transferred into the setter)

engrosamaali91 replied:
Removed underscore:

self.crop_position = crop_position

dkuegler commented on FastSurferCNN/models/interpolation_layer.py:
```diff
@@ -490,29 +492,31 @@ def _interpolate(
         target_shape : _T.Optional[_T.Sequence[int]]
             Target tensor size for after this module,
             not including batchsize and channels.
         interpolation_mode : str
```
`interpolation_mode : str, default="nearest"`

According to the numpy docstring code style this is the format for default values.
in some places in the code, we use "optional" to indicate a default value, that should also be changed.

dkuegler replied:
i.e. this change should be propagated throughout the code.

dkuegler commented on FastSurferCNN/models/losses.py:
```diff
@@ -13,50 +13,52 @@
+from numbers import Real
+from typing import Optional, Tuple, Union
+
 # IMPORTS
```
#IMPORTS first

dkuegler commented on FastSurferCNN/models/losses.py:
```diff
@@ -13,50 +13,52 @@
+        target: Tensor,
+        weights: Optional[int] = None,
+        ignore_index: Optional[int] = None,
     ) -> float:
```
return type should be torch.Tensor

dkuegler commented on FastSurferCNN/models/losses.py:
```diff
@@ -92,29 +94,30 @@ def forward(
 
         Parameters
         ----------
         weight : Optional[Tensor]
```
`Tensor, optional`....
propagate this change throughout all code.

dkuegler commented on FastSurferCNN/models/networks.py:
```diff
@@ -13,18 +13,21 @@
 import yacs
 
-import FastSurferCNN.models.sub_module as sm
+# IMPORTS
```
#IMPORTS up top (everywhere)

dkuegler commented on FastSurferCNN/models/networks.py:
```diff
@@ -50,15 +53,16 @@ class FastSurferCNNBase(nn.Module):
         params : Dict
+            Parameters in dictionary format
 
         padded_size : int
```
int, default=256

dkuegler commented on FastSurferCNN/models/networks.py:
```diff
@@ -89,27 +93,27 @@ def __init__(self, params: Dict, padded_size: int = 256):
         x : Tensor
-            input image [N, C, H, W]
+            Input image [N, C, H, W] representing the input data.
         scale_factor : Optional[Tensor]
```
```
scale_factor : Tensor, optional
```
Also remove the default = None text (that is what "optional" means)

dkuegler commented on FastSurferCNN/quick_qc.py:
```diff
@@ -70,22 +69,22 @@ def options_parse():
-        [MISSING]
-    thres :
-        [MISSING]
+    asegdkt_segfile : -
```
type - use np.ndarray

also add to function declaration

dkuegler commented on FastSurferCNN/quick_qc.py:
```diff
@@ -70,22 +69,22 @@ def options_parse():
-        [MISSING]
+    asegdkt_segfile : -
+        [MISSING].
+    voxvol : -
```
type float

dkuegler commented on FastSurferCNN/quick_qc.py:
```diff
@@ -70,22 +69,22 @@ def options_parse():
+        [MISSING].
+    voxvol : -
+        [MISSING].
+    thres : -
```
type float

dkuegler commented on FastSurferCNN/quick_qc.py:
```diff
@@ -101,27 +100,27 @@ def check_volume(asegdkt_segfile, voxvol, thres=0.70):
+        Segmentation array.
     region_labels : Dict
-        dict whose values correspond to the desired region's labels (Default value = VENT_LABELS)
+        Dict whose values correspond to the desired region's labels (Default value = VENT_LABELS).
```
?? default value not a dictionary?

dkuegler replied:
you need to describe what VENT_LABELS is (see VENT_LABELS definition up top)

engrosamaali91 replied:
**VENT_LABELS is indeed a dictionary:**

VENT_LABELS = {
    "Left-Lateral-Ventricle": 4,
    "Right-Lateral-Ventricle": 43,
    "Left-choroid-plexus": 31,
    "Right-choroid-plexus": 63,
}


**Described as follows:**

    region_labels : Dict
                Dictionary whose values correspond to the desired region's labels (Default value = VENT_LABELS).
                VENT_LABELS is a dictionary containing labels for different regions related to the ventricles,
                such as "Left-Lateral-Ventricle", "Right-Lateral-Ventricle", etc., along with their 
                corresponding numeric values.

dkuegler commented on FastSurferCNN/quick_qc.py:
```diff
@@ -101,27 +100,27 @@ def check_volume(asegdkt_segfile, voxvol, thres=0.70):
+        Dict whose values correspond to the desired region's labels (Default value = VENT_LABELS).
     bg_label : int
-        (Default value = BG_LABEL)
+        (Default value = BG_LABEL).
```
replace BG_LABEL with its value (e.g. make it a format str `f" bla {BG_LABEL} bla" `

dkuegler replied:
also should be in line above `bg_label : int, default={BG_LABEL}`

dkuegler commented on FastSurferCNN/run_prediction.py:
```diff
@@ -104,22 +110,22 @@ def args2cfg(args: argparse.Namespace) -> Tuple[yacs.config.CfgNode, yacs.config
 
 
 def removesuffix(string: str, suffix: str) -> str:
```
this function can be removed, and in places where it is called, replace `removesuffix(string, suffix)` with `string.removesuffix(suffix)`

engrosamaali91 replied:
I guess this is a standalone function, this function is not being called anywhere in run_prediction.py.

dkuegler commented on FastSurferCNN/utils/arg_types.py:
```diff
@@ -22,22 +22,25 @@
     Returns
     -------
-    [MISSING]
+    str or float
```
:+1:

dkuegler commented on FastSurferCNN/utils/arg_types.py:
```diff
@@ -22,22 +22,25 @@
     Raises
     ------
     argparse.ArgumentTypeError
         An error from creating or using an argument. Additionally, vox_sizes may be 'min'.
```
Should describe, when an error is raised (=if the arguemnt is not "min", "auto" or convertible to a float between 0 and 1)

dkuegler commented on FastSurferCNN/utils/arg_types.py:
```diff
@@ -50,17 +53,20 @@ def vox_size(a: str) -> VoxSizeOption:
+    float or None
+        If `a` is a valid float between 0 and 1, return the float value.
+        If `a` is 'none' or 'infinity', return None.
+        Otherwise, raise an argparse.ArgumentTypeError.
```
:+1:

dkuegler replied:
The raise description should be in a section called Raises

```
Raises
---------
argparse.ArgumentTypeError
    If `a` is neither a float between 0 and 1 or....
```

dkuegler commented on FastSurferCNN/utils/arg_types.py:
```diff
@@ -72,22 +78,23 @@ def float_gt_zero_and_le_one(a: str) -> Optional[float]:
     ----------
     a : str
-        datatype
+        Datatype.
```
Datatype descriptor

dkuegler commented on FastSurferCNN/utils/arg_types.py:
```diff
@@ -72,22 +78,23 @@ def float_gt_zero_and_le_one(a: str) -> Optional[float]:
     argparse.ArgumentTypeError
-        Invalid dtype
-
+        Invalid dtype.
```
Add 

```
See Also 
-----------
numpy.dtype

```

dkuegler commented on FastSurferCNN/utils/arg_types.py:
```diff
@@ -157,18 +164,18 @@ def int_ge_zero(value) -> int:
     val : str
-        A string of the value without quoting with '''
-
+        A string of the value without quoting with '''.
```
triple-quote looks wrong.

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -178,31 +180,30 @@ def pipeline(
-        [MISSING]
-
+    Iterator[Tuple[_Ti, _T]]
+        Iterator yielding input elements and corresponding results.
```
... yielding elements from `iterable` and corresponding results from `func(element)`

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -224,46 +225,46 @@ def pipeline(
-        iterable
+        Iterable.
 
     Yields
```
This is function has the same signature as the pipeline function above. Please change it so both use Returns or Yields.

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -224,46 +225,46 @@ def pipeline(
      element : _Ti
-        elements
+        Elements
     _T
```
the second return is `func(element)`

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -224,46 +225,46 @@ def pipeline(
     pool : Executor
-        [MISSING]
+        [MISSING].
     func : Callable[[_Ti], _T]
```
just use callable here

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -224,46 +225,46 @@ def pipeline(
     func : Callable[[_Ti], _T]
-        function to use
+        Function to use.
     iterable : Iterable[_Ti]
```
just use type Iterable here

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -224,46 +225,46 @@ def pipeline(
         yield element, func(element)
 
 
 def removesuffix(string: str, suffix: str) -> str:
```
remove this function, replace occurances with string.removesuffix as described above in other file.

engrosamaali91 replied:
The function removesuffix((string: str, suffix: str) is called on line 1010 like this:

`os.path.basename(removesuffix(subject, self._remove_suffix))
`

Shall i replace it like this after removing the function?

`  os.path.basename(string.removesuffix(suffix))
`

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -303,24 +307,23 @@ def __init__(self, **kwargs):
         ----------
         filepath : str
-            abs path to the file or name of the file
+            Abs path to the file or name of the file.
```
Absolute path

dkuegler commented on FastSurferCNN/utils/common.py:
```diff
@@ -605,84 +610,87 @@ def main_segfile(self) -> str:
         -------
-            Value of the attribute
-
+        AttributeError
```
Should be a a Raises section, wrong value for a Return section

dkuegler reviewed (changes requested):
I looked at files up to but not including FastSurferCNN/utils/mapper.py

dkuegler commented on doc/api/FastSurferCNN.utils.rst:
```diff
@@ -0,0 +1,24 @@
+API Utils References
```
should this not be "utils"? lowercase?

engrosamaali91 replied:
Changed.

dkuegler commented on doc/api/index.rst:
```diff
@@ -0,0 +1,12 @@
+API
```
FastSurfer API

dkuegler reviewed (changes requested):
minor changes after rebase

engrosamaali91 committed:
Implemented requested changes

dkuegler commented on FastSurferCNN/models/interpolation_layer.py:
```diff
@@ -416,7 +409,7 @@ def __init__(
         self._N = 2
         super(Zoom2d, self).__init__(target_shape, interpolation_mode)
-        self._crop_position = crop_position
+        self.crop_position = crop_position
```
I dont see the implementation of the property crop_position

dkuegler committed:
Changes

arg_types.py
- update unquote_str docstring

common.py
- remove the obsolete removesuffix function

run_prediction.py
- remove the obsolete removesuffix function

quick_qc.py
- Fix the docstring of get_region_by_intersection_mask

generate_hdf5.py
- replace seg-mask with sag_mask

interpolation_layer.py
- Update of the crop_position property

common.py
- Update the docstrings of pipeline and iterate
